README.md
# simplified.js
A javascript library which aims to simplify some javascript functions. In the long run, it also aims to be a reference for javascript developers, which want to learn typical JS-tasks the native way, instead of using frameworks like jQuery. The source code is all one file, which makes it easy to search up a function, and see how that specific function works under the hood. Most of all, this project was started for my own learning sake.

## Installation
It is important that you important simplified.js before importing your own script.
```
<script src="https://cdn.rawgit.com/markusvhagen/simplified.js/master/src/simplified.js"></script>
<script src="myScript.js"></script>
```
If you want, you can also download the code itself by going to [the raw source code](https://raw.githubusercontent.com/markusvhagen/simplified.js/master/src/simplified.js) and clicking `CTRL + S`.

## Functions in simplified.js
All functions in simplified.js has a underscore `_` in front. This is to prevent functions in simplifed.js from overwriting functions in your own .js-file.

### `_getElById(id)`
Returns an element as an object. Takes string as argument. Equivalent of `document.getElementById("myId")`.

### `_getContent(el)`
Returns innerHTML of an element. Takes an object (for example by `_getElById()`) as argument. Equivalent of `document.getElementById("myId").innerHTML`.

### `_setContent(el, content)`
Sets innerHTML of a given element. Takes an object as the first argument, and a string as the second argument.

### `_show(el)`
Sets display of element to `block`. Takes an object as argument.

### `_hide(el)`
Sets display of element to `none`. Takes an object as argument.

### `_getCSSValue(el, property)`
Returns a value of an certain CSS-property of an element. Takes an object as the first argument, and string as the second argument.

### `_setCSSValue(el, property, val)`
Sets the value of a given CSS-property of an element. Takes object as first argument and string as the second and third argument.

### `_fadeIn(el, timeInput)`
Fades in an element, and sets display-property of element to block. Takes object as the first argument and number as the second argument. You can read more about the `timeInput`-argument in the chapter [Timing with simplified.js](#timing-with-simplified.js).

### `_fadeOut(el, timeInput)`
Fades out an element, and sets display-property of element to none. Takes object as the first argument and number as the second argument. You can read more about the `timeInput`-argument in the chapter [Timing with simplified.js](#timing-with-simplified.js).

### `_scrollToTop(timeInput)`
Scrolls up to the top of the page smoothly in given time. Takes a number as argument. You can read more about the `timeInput`-argument in the chapter [Timing with simplified.js](#timing-with-simplified.js).

### `_scrollTo(el, timeInput)`
Scrolls smoothly to a given element in given time. Takes an object as the first argument, and a number as the second argument. You can read more about the `timeInput`-argument in the chapter [Timing with simplified.js](#timing-with-simplified.js).

### `_dynamicIntegerCounter(el, start, stop, timeInput)`
Will make an animation for a given element of couting from a given number (`start`) to a given number (`stop`) in a given time. Takes object as the first argument, and a number as the second, third and fourth argument. You can read more about the `timeInput`-argument in the chapter [Timing with simplified.js](#timing-with-simplified.js).

### `_indexOf(inArray, searchFor)`
Given a sorted array, it will return the index of `searchFor`. If `searchFor` is not in the array, it will return `undefined`. Does not worked with mixed arrays, just arrays only existing of numbers or only existing of strings. Takes an array as first argument, and string or number as the second argument. Under the hood, `_indexOf` will do a binary search, which is faster than native JS (`array.indexOf()`), because it searches linearly.

## Timing with simplified.js
Functions that can be timed has parameters for it, meaning that you can pass arguments regarding the timing of a function. For now, simplified.js only support linear timing, meaning the same speed of the animation in all phases. Timing is passed as an argument in milliseconds. If no argument is passed, the timing will be set to `default`, which is `400ms`.

In other words, you may pass an argument for time in the following ways:
* As an integer with milliseconds.
* As either `"slow"` (800 ms) or  `"fast"` (200 ms).  
* Passing no argument, and setting the speed to default.

## Validating arguments with simplified.js
Simplified.js uses the object validation for validating function arguments. The validation-object has two functions; `validateSingleArgument` and `validateMultipleArguments`. The object was originally made to validate arguments passed to simplified.js-function, but you can also use it. It works as follows;

```javascript
function myFunc(arg1, arg2) {
    if(validation.validateMultipleArguments(arg1, arg2, "string", "number")) {
        console.log("I will only print if arg1 is a string and arg2 is a number");
    }
}
```


changelog.md
## First entry - version 0.1 &nbsp;&nbsp;&nbsp;&nbsp; (04.06.2018)
It was probably about time that a changelog was made. Because there are no other versions to compare the current version to, I will just sum up what simplified.js can do for you so far.
* Get a DOM-Object by its id.
* Get/Set innerHTML of an element.
* Show/hide an element.
* Get/Set CSS-values of an element.
* Fade in/Fade out an element.
* Scroll to top of page or scroll to a specific element on a page.
* Make a dynamic integer counter.
* Search for an element in a sorted array (binary search).
* Validating arguments to functions.
* Has flexible timing-options for fade and scrolling functions.


tests/test.js
var show = _getElById("show");
var test = _getElById("test"); // NOT USE THIS TO TEST SCROLL; SCROLL DOES NOT WORK ON DISPLAY:NONE ELEMENTS.
var hide = _getElById("hide");
var scrollTop = _getElById("test_scroll");
var scrollToHere = _getElById("test_scrollTo");
var styleBtn = _getElById("styleBtn");
var counter = _getElById("counter");
var activateCounter = _getElById("activate_counter");
var scrollDown = _getElById("test_scrollDown");
var scrollDownToHere = _getElById("test_scrollDownTo");

show.onclick = function() {
    _fadeIn(test, 2000);
}

hide.onclick = function() {
    _fadeOut(test, 2000);
}

scrollTop.onclick = function() {
    _scrollTo(scrollToHere, 2000);
}

styleBtn.onclick = function() {
    _setCSSValue(hide, "color", "yellow");
    _setContent(hide, "I like fish :D");
}

activateCounter.onclick = function() {
    _dynamicIntegerCounter(counter, 0, 1000, "fast");
}

scrollDown.onclick = function() {
    _scrollTo(scrollDownToHere, 2000);
}

window.onload = function() {
    console.log(_indexOf([3,4,5,6,7,8], 4));
    console.log(_indexOf(["a", "ab", "cd3", "de3", "es1"], "es1"));
}


src/simplified.js
/*
    Simplified.js
    Licensed under the MIT license.
*/

"use strict";

/* All functions have "_" in front of it. This is to prevent it from
overwriting any other functions in another file */

/* === OBJECTS === */

let _linearSpeed = {
    default: 400,
    slow: 800,
    fast: 200,
    decideSpeed: function (stringInput) {
        if (stringInput === undefined) {
            return this.default;
        }
        // Only return if number is not integer.
        else if (!Number.isInteger(stringInput)) {
            if (stringInput == "slow") {
                return this.slow;
            }
            else if (stringInput == "fast") {
                return this.fast;
            }
        }
        else {
            return stringInput;
        }
    }
};

let _validation = {
    validateSingleArgument: function (argument, parameterType) {
        if (typeof(argument) === parameterType) {
            return true;
        }
        else {
            return false;
        }
    },

    validateMultipleArguments: function() {
        // Arguments is an object which is in JS natively.
        let numberOfArguments = arguments.length;
        let validationSuccess = true;

        // Only proceed if numberOfArguments is even, meaning that every argument for input-argument
        // has a corresponding argument for which type this argument should be.
        if (numberOfArguments % 2 == 0) {
            for (var i = 0; i <= ((numberOfArguments/2)-1); i++) {
                let currentCorrespondingArgument = arguments[(numberOfArguments/2)+i];
                if (!(typeof arguments[i] == currentCorrespondingArgument)) {
                    validationSuccess = false;
                    break;
                }
            }
        }

        else {
            return false;
        }
        return validationSuccess ? true : false;
    }

};

/* ==== SELECTORS ==== */

// Argument must be id of element
function _getElById(id) {
    if (_validation.validateSingleArgument(id, "string")) {
        return document.getElementById(id);
    }
}

/* ==== METHODS & EVENTS ==== */

function _getContent(el) {
    if (_validation.validateSingleArgument(el, "object")) {
        return el.innerHTML;
    }
}

function _setContent(el, content) {
    if (_validation.validateMultipleArguments(el, content, "object", "string")) {
        el.innerHTML = content;
    }
}

function _show(el) {
    if (_validation.validateSingleArgument(el, "object")) {
        if (el.style.display == "block") {
            return false;
        }
        else {
            el.style.display = "block";
        }
    }
}

function _hide(el) {
    if (_validation.validateSingleArgument(el, "object")) {
        if (el.style.display == "none") {
            return false;
        }
        else {
            el.style.display = "none";
        }
    }
}

function _getCSSValue(el, property) {
    if (_validation.validateMultipleArguments(el, property, "object", "string")) {
        let propVal = window.getComputedStyle(el).getPropertyValue(property);
        return propVal;
    }
}

function _setCSSValue(el, property, val) {
    if (_validation.validateMultipleArguments(el, property, val, "object", "string", "string")) {
        el.style[property] = val;  // Does not work when trying to access with el.style.property. Therefore el.style[property].
    }
}

function _fadeIn(el, timeInput) {
    if (_validation.validateMultipleArguments(el, "object")) {
        let time = _linearSpeed.decideSpeed(timeInput);
        let loopRounds = Math.ceil(time/16);
        let elOpacity = 0;
        let numberLoopRoundsNow = 1;
        el.style.opacity = elOpacity;
        el.style.display = "block";
        var fadeInInterval = setInterval(function() {
            if (numberLoopRoundsNow >= loopRounds) {
                clearInterval(fadeInInterval);
            }
            el.style.opacity = numberLoopRoundsNow/loopRounds;
            numberLoopRoundsNow++;
        }, 16);
    }
}

function _fadeOut(el, timeInput) {
    if (_validation.validateMultipleArguments(el, "object")) {
        let time = _linearSpeed.decideSpeed(timeInput);
        let loopRoundsTotal = Math.ceil(time/16);
        let elOpacity = 1;
        let loopRoundsNow = Math.ceil(time/16);
        el.style.opacity = 1;
        var fadeOutInterval = setInterval(function() {
            if (loopRoundsNow <= 0) {
                clearInterval(fadeOutInterval);
                el.style.display = "none";
            }
            el.style.opacity = loopRoundsNow/loopRoundsTotal;
            loopRoundsNow--;
        }, 16);
    }
}

function _scrollToTop(timeInput) {
    let time = _linearSpeed.decideSpeed(timeInput);
    let scrollPos = window.scrollY;
    if (scrollPos == 0) {
        return true;
    }
    else {
        let loopRoundsTotal = Math.ceil(time/16);
        let pixelMovePerLoop = Math.ceil(scrollPos/loopRoundsTotal);
        let loopRoundsNow = 0;
        var scrollTopInterval = setInterval(function() {
            if (document.body.scrollTop === 0) {
                clearInterval(scrollTopInterval);
            }
            scrollPos -= pixelMovePerLoop;
            window.scrollTo(0, scrollPos);
            loopRoundsNow++;
        }, 16);
    }
}

function _scrollTo(el, timeInput) {
    if (_validation.validateMultipleArguments(el, "object")) {
        let time = _linearSpeed.decideSpeed(timeInput);
        let scrollPos = window.scrollY;
        let scrollToPos = el.getBoundingClientRect().y - document.body.getBoundingClientRect().y;
        let loopRoundsTotal = Math.ceil(time/16);
        // If element has a lower Y-coordinate then starting scroll position.
        if (scrollPos > scrollToPos) {
            let pixelMovePerLoop = Math.abs(Math.ceil((scrollPos - scrollToPos)/loopRoundsTotal));
            var scrollToElementInterval = setInterval(function() {
                if (window.scrollY <= scrollToPos) {
                    clearInterval(scrollToElementInterval);
                }
                scrollPos -= pixelMovePerLoop;
                window.scrollTo(0, scrollPos);
            }, 16);
        }
        // If element has a higher Y-coordinate then starting scroll position.
        else if (scrollPos < scrollToPos) {
            let pixelMovePerLoop = Math.abs(Math.ceil((scrollPos - scrollToPos)/loopRoundsTotal));
            var scrollToElementInterval = setInterval(function() {
                if (window.scrollY >= scrollToPos) {
                    clearInterval(scrollToElementInterval);
                }
                scrollPos += pixelMovePerLoop;
                window.scrollTo(0, scrollPos);
            }, 16);
        }
    }
}

function _dynamicIntegerCounter(el, start, stop, timeInput) {
    if (_validation.validateMultipleArguments(el, start, stop, "object", "number", "number")) {
        let time = _linearSpeed.decideSpeed(timeInput);
        let difference = stop - start;
        let msPerStep = time / difference;
        let i = 0;

        var counterInterval = setInterval(function () {
            if (i == stop) {
                clearInterval(counterInterval);
            }
            el.innerHTML = i;
            i++;
        }, msPerStep);
    }
}

// Does NOT accept mixed arrays, just numbers or just strings
function _indexOf(inArray, searchFor) {
    if (typeof inArray[0] == "number") {
        var minIndex = 0;
        var maxIndex = inArray.length - 1;
        var midIndex;

        while (maxIndex >= minIndex) {
            midIndex = Math.floor((minIndex+maxIndex)/2);

            if (searchFor < inArray[midIndex]) {
                maxIndex = midIndex - 1;
            }
            else if (searchFor > inArray[midIndex]) {
                minIndex = midIndex + 1;
            }
            else if (searchFor == inArray[midIndex]) {
                return midIndex;
            }
            else {
                return undefined;
            }
        }
    }

    else if (typeof inArray[0] == "string") {
        var minIndex = 0;
        var maxIndex = inArray.length - 1;
        var midIndex;

        while (maxIndex >= minIndex) {
            midIndex = Math.floor((minIndex+maxIndex)/2);

            // If searchFor has a lower index
            if (searchFor.localeCompare(inArray[midIndex]) == -1) {
                maxIndex = midIndex - 1;
            }
            // If searchFor has a higher index
            else if (searchFor.localeCompare(inArray[midIndex]) == 1) {
                minIndex = midIndex + 1;
            }
            else if (searchFor.localeCompare(inArray[midIndex]) == 0) {
                return midIndex;
            }
            else {
                return undefined;
            }
        }
    }
}
